### PR TITLE
add `OpInfo` for `torch.nn.functional.dropout`

### PR DESCRIPTION
**Pull request by pmeier**
Addresses facebookresearch/functorch#78.

cc @albanD @mruberry @jbschlosser @walterddr

### CODE REVIEW

**Comment by pmeier** [comment]
Failures like [this](https://app.circleci.com/pipelines/github/pytorch/pytorch/357044/workflows/7c373aaa-15ad-4afa-8b2c-ce7f0de955f8/jobs/15054205/tests#failed-test-2) look like legitimate bugs in the op:

```
torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor(-8101425.5804, device='cuda:0')
analytical:tensor(0.9649, device='cuda:0')

The above quantities relating the numerical and analytical jacobians are computed 
in fast mode. See: https://github.com/pytorch/pytorch/issues/53876 for more background 
about fast mode. Below, we recompute numerical and analytical jacobians in slow mode:

Numerical:
 tensor([[-6.3614e+03,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
        [ 0.0000e+00, -3.1640e+03,  7.6229e+03,  0.0000e+00,  0.0000e+00],
        [ 6.3624e+03,  0.0000e+00,  2.0000e+00,  4.4501e+03,  0.0000e+00],
        [-6.3624e+03,  3.1650e+03, -7.6229e+03, -4.4491e+03,  0.0000e+00],
        [ 0.0000e+00,  3.1650e+03,  7.6229e+03,  0.0000e+00,  2.2878e+03]],
       device='cuda:0')
Analytical:
tensor([[0., 0., 0., 0., 0.],
        [0., 2., 0., 0., 0.],
        [0., 0., 2., 0., 0.],
        [0., 0., 0., 2., 0.],
        [0., 0., 0., 0., 0.]], device='cuda:0')

The max per-element difference (slow mode) is: 7622.938523997853.
```

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -7875,6 +7887,14 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    decorators=(toleranceOverride({torch.float32: tol(atol=0, rtol=4e-6), }),),
                    dtypes=all_types_and(torch.bool),
                    safe_casts_outputs=True),
+    OpInfo(
+        "nn.functional.dropout",
+        ref=_NOTHING,
```
Comment (pmeier) [inline torch/testing/_internal/common_methods_invocations.py:7892]: Due to the random behavior of dropout, I don't think it is possible to have a black-box reference in `numpy`.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -7875,6 +7887,14 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+        supports_out=False,
+        dtypes=floating_types(),
+        sample_inputs_func=sample_inputs_dropout,
+        inplace_variant=partial(torch.nn.functional.dropout, inplace=True),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7896]: Nice inplace variant

**Comment by mruberry** [comment]
This one has a ton of test failures, which seem real, possibly because our tests aren't consistently resetting randomness across calls to our operators?

This will probably require fixing tests to support. Maybe @krshrimali would be interested in helping with that?

**Comment by kshitij12345** [comment]
Thanks to offline discussion with @krshrimali , we realized that adding gradcheck would be hard (maybe that is why we didn't see any gradcheck tests for dropout). This is because of the random mask generated by dropout on each call. Due to this output of function is different while computing the numerical gradient and analytical gradient (hence the gradcheck failures).
Not sure if there is an easy way around this

cc: @albanD @soulitzer 

```python
import torch
torch.manual_seed(42)
t = torch.ones(5, dtype=torch.float64, requires_grad=True)

# Works if the mask stays the same.
p = 0.5
m = torch.ones_like(t).bernoulli_(p) / (1 - p)

class Dropout(torch.autograd.Function):
    def __init__(self, p):
        self.p = p

    @staticmethod
    def forward(ctx, i):
        # Doesn't work if the mask changes on every
        # invocation.
        # m = torch.ones_like(t).bernoulli_(p) / (1 - p)
        result = i * m
        ctx.save_for_backward(m)
        return result

    @staticmethod
    def backward(ctx, grad_output):
        m, = ctx.saved_tensors
        return grad_output * m

torch.autograd.gradcheck(lambda t: Dropout.apply(t), t)

```

**Comment by albanD** [comment]
Does changing the gradcheck call to
```python
```

**Comment by kshitij12345** [comment]
@albanD that does work! Thanks. That should solve the problem!

cc: @krshrimali

**Review by zou3519 — COMMENTED** [review]
Test failures seem real

**Comment by kshitij12345** [comment]
@krshrimali You will have to wrap `dropout` op and the variants with a seed prior to calling it.

The eager variant test and other tests seem to be failing because the dropped elements differ between the call to variants.

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -8286,6 +8304,30 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+            wrapper_set_seed(torch.nn.functional.dropout, input, *args, **kwargs),
+        ref=_NOTHING,
+        dtypes=floating_types(),
+        dtypesIfCUDA=floating_types_and(torch.float16, torch.bfloat16),
+        skips=(
+            # Probably because we have used lambda for the op here
+            # AssertionError: JIT Test does not execute any logic
+            SkipInfo('TestJit', 'test_variant_consistency_jit'),
+            # inplace variant dispatches to dropout kernel, while on CUDA
+            # the op dispatches to _fused_dropout (with a few more conditions)
```
Comment (krshrimali) [inline torch/testing/_internal/common_methods_invocations.py:8319]: Please see: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Dropout.cpp#L24 and https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Dropout.cpp#L88 for more context.

**Comment by krshrimali** [comment]
Hi, @zou3519 and @mruberry - This PR is now ready for your review, apologies for the delay. PTAL, whenever you find the time, thanks!

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -8959,6 +8977,30 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    decorators=(toleranceOverride({torch.float32: tol(atol=0, rtol=4e-6), }),),
                    dtypes=all_types_and(torch.bool),
                    safe_casts_outputs=True),
+    OpInfo(
+        "nn.functional.dropout",
+        op=lambda input, *args, **kwargs:
+            wrapper_set_seed(torch.nn.functional.dropout, input, *args, **kwargs),
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:8983]: cc @mruberry is this okay? The [documentation for `op`](https://github.com/pytorch/pytorch/blob/87d8ab6e50314f72ea6730250904fe7ff073cc21/torch/testing/_internal/common_methods_invocations.py#L469) says "the function variant of the operation, populated as torch.<name> if None". The wrapper is definitely not "the function variant of the operation".

If `op` should not be a wrapper, then we could consider something like adding a new flag to OpInfo. Something like `use_seed=None`.

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:8983]: If we get to n=3 on this particular issue then I agree it's interesting to think about adding architecture to support it, but a wrapper like this seems convenient for now

File: test/test_fx_experimental.py [hunk 5]
```diff
@@ -1460,6 +1460,7 @@ def test_normalize_operator_exhaustive(self, device, dtype, op):
         # Sorted and one entry on each line to minimize merge conflicts.
         op_skip = {
             "contiguous",
+            "nn.functional.dropout",
```
[the remark this answers is not in the document]
Comment (krshrimali) [reply test/test_fx_experimental.py]: Done, sorry didn't notice that this is alphabetically sorted. :)

**Comment by krshrimali** [comment]
Hi, @zou3519 and @mruberry - can you please take a look at this PR whenever you find the time? Thanks!

**Review by zou3519 — APPROVED** [review]
This LGTM. @krshrimali could you rebase this please?

**Comment by krshrimali** [comment]
> This LGTM. @krshrimali could you rebase this please?

Thanks, @zou3519 for taking a look! Have rebased this :) Appreciate the reviews and patience with this PR.

**Comment by zou3519** [comment]
> > This LGTM. @krshrimali could you rebase this please?
> 
> Thanks, @zou3519 for taking a look! Have rebased this :) Appreciate the reviews and patience with this PR.

Thank you for the contribution and I appreciate your patience! I'm going to set myself a new goal to be more responsive to review requests.

**Pull request closed without merging**